client/keystore/src/lib.rs

```diff
@@ -276,8 +277,83 @@ impl Store {
 }
 
 impl BareCryptoStore for Store {
+	fn supported_keys(
+		&self,
+		id: KeyTypeId,
+		keys: Vec<CryptoTypePublicPair>
+	) -> std::result::Result<Vec<CryptoTypePublicPair>, String> {
+		let ed25519_existing_keys = self
+			.public_keys_by_type::<ed25519::Public>(id)
+			.map_err(|e| e.to_string())?;
+		let ed25519_existing_keys: Vec<Vec<u8>>  = ed25519_existing_keys
+			.iter()
+			.map(|k| k.to_raw_vec())
+			.collect();
+
+		let sr25519_existing_keys= self
+			.public_keys_by_type::<sr25519::Public>(id)
+			.map_err(|e| e.to_string())?;
+		let sr25519_existing_keys: Vec<Vec<u8>> = sr25519_existing_keys
+			.iter()
+			.map(|k| k.to_raw_vec())
+			.collect();
+
+		Ok(keys.iter().filter_map(|k| {
+			match k.0 {
+				sr25519::SR25519_CRYPTO_ID if sr25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
```

Comment:
This is pretty much quadratic complexity. I think this should be optimized into `Set` intersection. Currently if you pass a long list of `keys` and you have nothing in the keystore it's still linear, as we have to traverse every key, while having less keys in the keystore is a much more likely case - i.e. you pass a list of `1024` validators and you are just one of them.

Reply:
Given that we have `fn keys` below, let's try this:
```
let wanted_keys = keys.into_iter().collect::<HashSet<_>>();
let got_keys = self.keys()?.into_iter().collect::<HashSet<_>>();

Ok(wanted_keys.intersection(&got_keys).collect())
```